### PR TITLE
Add TLS support for kvrocks connections

### LINKED ISSUE #147: Any plan to support SSL?
libevent provides ssl with bufferevent_ssl.h and openssl. I guess is not hard to support ssl.  Is there any plan to support ssl?

### PR DESCRIPTION
**Pull request by PragmaTwice**
We bring TLS support for kvrocks, referring to [redis TLS](https://redis.io/docs/manual/security/encryption/).

TODO:
- [x] conditional compilation
- [x] more friendly error messages
- [x] more options for SSL cert/key/cacert and ciphers
- [x] thread safety for openssl and eventbuffer
- [x] hot-reload tls options
- [x] tests

temporarily not in plan (of this PR):
- tls for replication

---

**Why do we use shared library of OpenSSL and do not pin the version of OpenSSL?**

OpenSSL contains many cryptographic algorithms and implementations of the TLS protocol, which are security-critical. We hope that users will be able to quickly update the OpenSSL that kvrocks depends on to a new version in the event of a **serious security vulnerability** in OpenSSL (In fact OpenSSL has had several serious security incidents). If static linking is used, users may need to modify the cmake file and recompile, and kvrocks developers also need to always pay attention to the security vulnerabilities of OpenSSL and maintain version of cmake dependencies, which is very troublesome.

---

Supported options:

- tls_port
- tls_cert_file
- tls_key_file
- tls_key_file_pass
- tls_ca_cert_file
- tls_ca_cert_dir
- tls_auth_clients
- tls_prefer_server_ciphers
- tls_ciphers
- tls_ciphersuites
- tls_protocols
- tls_session_caching
- tls_session_cache_size
- tls_session_cache_timeout

### CODE REVIEW

**Comment by PragmaTwice** [comment]
Hi all, this PR is now ready for review!

There is an error in TCL TLS test only in macOS, and I think I test a lot to ensure it works well in linux(centos or archlinux, openssl 1.0 or 1.1, tls v1 v1.1 v1.2 or v1.3).
I will try to figure out the error in mac, and I will be very appreciated if you help me, since I have no Mac device.

cc @git-hulk @ShooterIT @tisonkun @caipengbo

**Comment by git-hulk** [comment]
> Hi all, this PR is now ready for review!
> 
> There is an error in TCL TLS test only in macOS, and I think I test a lot to ensure it works well in linux(centos or archlinux, openssl 1.0 or 1.1, tls v1 v1.1 v1.2 or v1.3). I will try to figure out the error in mac, and I will be very appreciated if you help me, since I have no Mac device.
> 
> cc @git-hulk @ShooterIT @tisonkun @caipengbo

Thanks @PragmaTwice, I will try to locate macOS issue, not sure whether it caused by macOS ssl configured to forbid the older version like 1.2

**Comment by PragmaTwice** [comment]
> > Hi all, this PR is now ready for review!
> > There is an error in TCL TLS test only in macOS, and I think I test a lot to ensure it works well in linux(centos or archlinux, openssl 1.0 or 1.1, tls v1 v1.1 v1.2 or v1.3). I will try to figure out the error in mac, and I will be very appreciated if you help me, since I have no Mac device.
> > cc @git-hulk @ShooterIT @tisonkun @caipengbo
> 
> Thanks @PragmaTwice, I will try to locate macOS issue, not sure whether it caused by macOS ssl configured to forbid the older version like 1.2

So thanks! The version of openssl in mac is probably 3.0, and you can check here to see the error:

https://github.com/apache/incubator-kvrocks/runs/8110693005?check_suite_focus=true

To reproduce the error, you can just follow `kvrocks.yml`:
```
            git clone https://github.com/jsha/minica
            cd minica && go build && cd ..
            ./minica/minica --domains localhost
            mkdir -p tests/tcl/tests/tls
            cp localhost/cert.pem tests/tcl/tests/tls/server.crt
            cp localhost/key.pem tests/tcl/tests/tls/server.key
            cp localhost/cert.pem tests/tcl/tests/tls/client.crt
            cp localhost/key.pem tests/tcl/tests/tls/client.key
            cp minica.pem tests/tcl/tests/tls/ca.crt
            ./x.py test tcl build --tls --single unit/tls --single unit/command --dont-clean
```

**Comment by git-hulk** [comment]
@PragmaTwice Yes, I took a look at openssl ciphers on my side and it only supported TLSv1.2 and SSLv3, but not sure what's the minica cert default cipher. Will take time to dive more recently.

**Comment by PragmaTwice** [comment]
> @PragmaTwice Yes, I took a look at openssl ciphers on my side and it only supported TLSv1.2 and SSLv3, but not sure what's the minica cert default cipher. Will take time to dive more recently.

Thanks! That is weird. Is the openssl system built-in, or installed from brew?

**Comment by git-hulk** [comment]
> > @PragmaTwice Yes, I took a look at openssl ciphers on my side and it only supported TLSv1.2 and SSLv3, but not sure what's the minica cert default cipher. Will take time to dive more recently.
> 
> Thanks! That is weird. Is the openssl system built-in, or installed from brew?

Installed from HomeBrew

File: README.md
Comment (git-hulk) [inline README.md:79]:
```suggestion
brew install autoconf automake libtool cmake openssl

# Please force linking the openssl if still can't find after installing openssl
brew link --force openssl
```

**Comment by PragmaTwice** [comment]
> > > @PragmaTwice Yes, I took a look at openssl ciphers on my side and it only supported TLSv1.2 and SSLv3, but not sure what's the minica cert default cipher. Will take time to dive more recently.
> > 
> > 
> > Thanks! That is weird. Is the openssl system built-in, or installed from brew?
> 
> Installed from HomeBrew

Thanks. I am now trying to configure more available ciphers and protocols to workaround it.

**Comment by git-hulk** [comment]
> > > > @PragmaTwice Yes, I took a look at openssl ciphers on my side and it only supported TLSv1.2 and SSLv3, but not sure what's the minica cert default cipher. Will take time to dive more recently.
> > > 
> > > 
> > > Thanks! That is weird. Is the openssl system built-in, or installed from brew?
> > 
> > 
> > Installed from HomeBrew
> 
> Thanks. I am now trying to configure more available ciphers and protocols to workaround it.

hmmmm... I tried it on my M1 machine, it works normally.

**Comment by git-hulk** [comment]
@PragmaTwice I didn't find out the reason but I manually test the tls on macOS and it works normally. Another piece of information is Redis also didn't enable tls test cases on macOS, can see the [redis daily ci](https://github.com/redis/redis/blob/unstable/.github/workflows/daily.yml). I guess maybe there's something wrong with `tcl-tls` since I failed to run tls test cases on my mac.

```
❯ ./runtest --tls --single unit/tls --single unit/info
Cleanup: may take some time... OK
Starting test server at port 21079
[ready]: 77784
Testing unit/tls
[ready]: 77785
Testing unit/info
[ready]: 77786
[ready]: 77783
[ready]: 77787
[ready]: 77788
[ready]: 77790
[ready]: 77789
[ready]: 77791
[ready]: 77792
[ready]: 77793
[ready]: 77795
[ready]: 77798
[ready]: 77796
[ready]: 77794
[ready]: 77797
--- LOG CONTENT ---
--- LOG CONTENT ---
[err]: Can't start the Redis server
CONFIGURATION:always-show-logo yes
notify-keyspace-events KEA
daemonize no
```

So I think we can disable tls test cases on mac as well. @ShooterIT @PragmaTwice @tisonkun  can correct me if I was wrong.

**Comment by PragmaTwice** [comment]
@git-hulk So thanks for your warm help! 

As discussed earlier, maybe tcl-tls is linked against the builtin openssl in mac, thus it behaves weirdly. I will temporarily disable the TCL TLS tests on MacOS so this PR can move on.

File: .github/workflows/kvrocks.yaml
Comment (git-hulk) [inline .github/workflows/kvrocks.yaml:135]: Can also use `brew link --force openssl`

File: src/worker.cc
Comment (git-hulk) [inline src/worker.cc:145]:
```suggestion
    SSL_free(ssl);
    evutil_closesocket(fd);
```

File: README.md
Comment (git-hulk) [inline README.md:79]:
```suggestion
brew install autoconf automake libtool cmake openssl
# Please force linking the openssl if still can't find it after installing openssl
brew link --force openssl
```

Comment (PragmaTwice) [reply README.md:79]: done

File: .github/workflows/kvrocks.yaml
Comment (PragmaTwice) [reply .github/workflows/kvrocks.yaml:135]: done

File: src/worker.cc
Comment (PragmaTwice) [reply src/worker.cc:145]: done

**Review by git-hulk — DISMISSED** [review]

File: .github/workflows/kvrocks.yaml
Comment (PragmaTwice) [inline .github/workflows/kvrocks.yaml]:
```suggestion
            ./x.py test tcl build --tls --single unit/tls --dont-clean
```

File: .github/workflows/kvrocks.yaml
Comment (git-hulk) [inline .github/workflows/kvrocks.yaml:68]: @PragmaTwice Why do we switch macOS to 11

Comment (PragmaTwice) [reply .github/workflows/kvrocks.yaml:68]: macos-11 is equal to macos-latest, referring to https://github.com/actions/runner-images#available-images

I used to try macos-12, but I reverted the change later.

**Comment by ShooterIT** [comment]
does replication support TLS?

**Comment by PragmaTwice** [comment]
> does replication support TLS?

Currently no. This PR does not provide any TLS client features.

It will be supported in subsequent PRs.

File: .github/workflows/kvrocks.yaml
Comment (git-hulk) [reply .github/workflows/kvrocks.yaml:68]: Okay, thanks

**Comment by git-hulk** [comment]
@ShooterIT Redis also separate the replication TLS, so I think it's a good idea to only implement server TLS in this PR. Even we don't implement the replication TLS unless someone required it.

Reference: https://github.com/redis/redis/blob/unstable/redis.conf#L254

File: src/config.cc
Comment (git-hulk) [inline src/config.cc]: We should not remove this line?

**Comment by git-hulk** [comment]
@ShooterIT @caipengbo Can take a look when you're free. This PR involved many files and it'll be conflicts frequently if it's pending too long.

File: src/config.cc
Comment (PragmaTwice) [reply src/config.cc]: It seems a bug in github, I do not remove this line. Just solved some conflict.

https://github.com/apache/incubator-kvrocks/pull/797/files#diff-76602be0b8dbf09757c540e3b4c474b477cac3bdf38ca318e6284cb34020cd2eR97-R105

Comment (git-hulk) [reply src/config.cc]: @PragmaTwice Yes, that's weird. It only can be seen in last commit, let ignore this.

**Review by git-hulk — APPROVED** [review]

**Comment by git-hulk** [comment]
@PragmaTwice I think we can merge this PR if no further discussion and other folks also can review and call a new discussion if they come back. cc @ShooterIT @caipengbo

**Comment by PragmaTwice** [comment]
> @PragmaTwice I think we can merge this PR if no further discussion and other folks also can review and call a new discussion if they come back. cc @ShooterIT @caipengbo

Great, I will merge it today.

**Comment by PragmaTwice** [comment]
Thanks all. Merging. Feel free to open issues or discussions if you have any problem or suggestion after merge.

**Pull request merged**
